From 0a591b0db3488974b8e94e7e68c0bef19330fb5d Mon Sep 17 00:00:00 2001
From: Ionut Vicovan <Ionut.Vicovan@nxp.com>
Date: Wed, 17 Feb 2021 16:52:13 +0200
Subject: [PATCH 78/84] pcie: Update PCIe mappings in S32 Gen1 device trees

- Fix overlaps in the memory map ranges.
- Put the config space at the very end with I/O space before.
This eliminates overlaps and permits best alignment for the memory
ranges.

Issue: ALB-6468

Signed-off-by: Heinz Wrobel <Heinz.Wrobel@nxp.com>
Signed-off-by: Ionut Vicovan <Ionut.Vicovan@nxp.com>
---
 .../boot/dts/freescale/fsl-s32-gen1.dtsi      | 56 ++++++++++++-------
 1 file changed, 36 insertions(+), 20 deletions(-)

diff --git a/arch/arm64/boot/dts/freescale/fsl-s32-gen1.dtsi b/arch/arm64/boot/dts/freescale/fsl-s32-gen1.dtsi
index 3f5ffbaaad9a..d0139a1e5e01 100644
--- a/arch/arm64/boot/dts/freescale/fsl-s32-gen1.dtsi
+++ b/arch/arm64/boot/dts/freescale/fsl-s32-gen1.dtsi
@@ -777,12 +777,16 @@
 	pcie0: pcie@40400000 {
 		compatible = "fsl,s32gen1-pcie";
 		reg = <0x00 0x40400000 0x0 0x00001000   /* dbi registers */
-		       0x00 0x40420000 0x0 0x00001000   /* dbi2 registers */
-		       0x00 0x40460000 0x0 0x00001000   /* atu registers */
-		       0x00 0x40470000 0x0 0x00001000   /* dma registers */
-		       0x00 0x40480000 0x0 0x00004000   /* ctrl registers */
-		       0x58 0x00000000 0x0 0x00002000   /* RC config space */
-		       0x58 0x40000000 0x0 0x40000000>; /* EP addr space, 1GB */
+			0x00 0x40420000 0x0 0x00001000   /* dbi2 registers */
+			0x00 0x40460000 0x0 0x00001000   /* atu registers */
+			0x00 0x40470000 0x0 0x00001000   /* dma registers */
+			/* ctrl (serdes) registers */
+			0x00 0x40480000 0x0 0x00004000
+			/* RC configuration space, 4KB each for cfg0 and cfg1
+			 * at the end of the outbound memory map
+			 */
+			0x5f 0xffffe000 0x0 0x00002000
+			0x58 0x00000000 0x0 0x40000000>; /* 1GB EP addr space */
 		reg-names = "dbi", "dbi2", "atu", "dma", "ctrl",
 				"config", "addr_space";
 		#address-cells = <3>;
@@ -790,10 +794,14 @@
 		device_type = "pci";
 		device_id = <0>;
 		ranges =
-		      /* downstream I/O */
-		      <0x81000000 0x0 0x00000000 0x58 0x00003000 0x0 0x00010000
-		      /* non-prefetchable memory */
-		      0x82000000 0x0 0x00013000 0x58 0x00013000 0x0 0x40000000>;
+			/* downstream I/O, 64KB and aligned naturally just
+			 * before the config space to minimize fragmentation
+			 */
+			<0x81000000 0x0 0x00000000 0x5f 0xfffe0000 0x0 0x00010000
+			/* non-prefetchable memory, with best case size and
+			 * alignment
+			 */
+			 0x82000000 0x0 0x00000000 0x58 0x00000000 0x7 0xfffe0000>;
 
 		num-lanes = <2>;
 		max-link-speed = <3>;
@@ -821,12 +829,16 @@
 	pcie1: pcie@44100000 {
 		compatible = "fsl,s32gen1-pcie";
 		reg = <0x00 0x44100000 0x0 0x00001000   /* dbi registers */
-		       0x00 0x44120000 0x0 0x00001000   /* dbi2 registers */
-		       0x00 0x44160000 0x0 0x00001000   /* atu registers */
-		       0x00 0x44170000 0x0 0x00001000   /* dma registers */
-		       0x00 0x44180000 0x0 0x00004000   /* ctrl registers */
-		       0x48 0x00000000 0x0 0x00002000   /* RC config space */
-		       0x48 0x40000000 0x0 0x40000000>; /* EP addr space, 1GB */
+			0x00 0x44120000 0x0 0x00001000   /* dbi2 registers */
+			0x00 0x44160000 0x0 0x00001000   /* atu registers */
+			0x00 0x44170000 0x0 0x00001000   /* dma registers */
+			/* ctrl (serdes) registers */
+			0x00 0x44180000 0x0 0x00004000
+			/* RC configuration space, 4KB each for cfg0 and cfg1
+			 * at the end of the outbound memory map
+			 */
+			0x48 0xffffe000 0x0 0x00002000
+			0x48 0x00000000 0x0 0x40000000>; /* 1GB EP addr space */
 		reg-names = "dbi", "dbi2", "atu", "dma", "ctrl",
 				"config", "addr_space";
 		#address-cells = <3>;
@@ -834,10 +846,14 @@
 		device_type = "pci";
 		device_id = <1>;
 		ranges =
-		      /* downstream I/O */
-		      <0x81000000 0x0 0x00000000 0x48 0x00003000 0x0 0x00010000
-		      /* non-prefetchable memory */
-		      0x82000000 0x0 0x00013000 0x48 0x00013000 0x0 0x40000000>;
+			/* downstream I/O, 64KB and aligned naturally just
+			 * before the config space to minimize fragmentation
+			 */
+			<0x81000000 0x0 0x00000000 0x4f 0xfffe0000 0x0 0x00010000
+			/* non-prefetchable memory, with best case size and
+			 * alignment
+			 */
+			 0x82000000 0x0 0x00000000 0x48 0x00000000 0x7 0xfffe0000>;
 
 		num-lanes = <1>;
 		max-link-speed = <3>;
-- 
2.17.1

